File: native/kmeans/CPU/kmeans.cpp
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <math.h>

#include "kmeans.h"
#include "point.h"

#define REPEAT 1

#define ITERATIONS 30

float distance(Point* p, Centroid* c) {
    float dx = p->x - c->x;
    float dy = p->y - c->y;
    return sqrtf(dx*dx + dy*dy);
}

void groupByCluster(
    Point* points,
    Centroid* centroids,
    size_t num_centroids, 
    size_t num_points
) {
#pragma omp parallel for simd
	for(size_t i0 = 0; i0 < num_points; i0++) {
		float minor_distance = -1.0;

		for (size_t i1 = 0; i1 < num_centroids; i1++) {
			float my_distance = distance(&points[i0], &centroids[i1]);
			if (minor_distance > my_distance || minor_distance == -1.0) {
				minor_distance = my_distance;
				points[i0].cluster = i1;
			}
		}
	}
}


void calCentroidsSum(
    Point* points, 
    Centroid* centroids,
    size_t num_centroids, 
    size_t num_points
) {
#pragma omp parallel for simd
    for(size_t i = 0; i < num_centroids; i++) {
        centroids[i].x_sum = 0.0;
        centroids[i].y_sum = 0.0;
        centroids[i].num_points = 0.0;
    }

    for(size_t i = 0; i < num_points; i++) {
        size_t ci = points[i].cluster;
        centroids[ci].x_sum += points[i].x;
        centroids[ci].y_sum += points[i].y;
        centroids[ci].num_points += 1;
    }
}


void updateCentroids(
    Centroid* centroids, 
    size_t num_centroids
) {
#pragma omp parallel for simd
	for(size_t i = 0; i < num_centroids; i++) {
	    if (centroids[i].num_points > 0) {
	        centroids[i].x = centroids[i].x_sum / centroids[i].num_points;
	        centroids[i].y = centroids[i].y_sum / centroids[i].num_points;
	    }
	}
}


void kmeans(
    Point* h_points,
    Centroid* h_centroids, 
    size_t num_points,
    size_t num_centroids
) {
    for(size_t i = 0; i < ITERATIONS; i++) {
        groupByCluster(
            h_points, 
            h_centroids,
            num_centroids, 
            num_points
        );
        
        calCentroidsSum(
            h_points, 
            h_centroids,
            num_centroids, 
            num_points
        );

        updateCentroids(
            h_centroids, 
            num_centroids
        );
    }
}

void printCentroids(
		    Centroid* centroids,
		    size_t NUMBER_OF_CENTROIDS
) {
    for (size_t i = 0; i < NUMBER_OF_CENTROIDS; i++) {
        printf("[x=%lf, y=%lf, x_sum=%lf, y_sum=%lf, num_points=%i]\n", 
               centroids[i].x, centroids[i].y, centroids[i].x_sum,
               centroids[i].y_sum, centroids[i].num_points);
    }

    printf("--------------------------------------------------\n");
}


void runKmeans(
    Point* points, 
    Centroid* centroids,
    size_t NUMBER_OF_POINTS,
    size_t NUMBER_OF_CENTROIDS
) {

    for (size_t i = 0; i < REPEAT; i++) {
        for (size_t ci = 0; ci < NUMBER_OF_CENTROIDS; ci++) {
            centroids[ci].x = points[ci].x;
            centroids[ci].y = points[ci].y;
        }

        kmeans(points, centroids, NUMBER_OF_POINTS, NUMBER_OF_CENTROIDS);
        
        //if (i + 1 == REPEAT) {
	//   printCentroids(centroids);
        //}
    }
}

File: numba/kmeans/GPU/kmeans.py
import dpctl
import numpy
import numba
import os

import base_kmeans
from device_selector import get_device_selector

REPEAT = 1
# defines total number of iterations for kmeans accuracy
ITERATIONS = 30

backend = os.getenv("NUMBA_BACKEND", "legacy")
if backend == "legacy":
    import numba as nb

    __njit = nb.jit(nopython=True, parallel=True, fastmath=True)
else:
    import numba_dpcomp as nb

    __njit = nb.jit(nopython=True, parallel=True, fastmath=True, enable_gpu_pipeline=True)


# determine the euclidean distance from the cluster center to each point
@__njit
def groupByCluster(arrayP, arrayPcluster, arrayC, num_points, num_centroids):
    # parallel for loop
    for i0 in numba.prange(num_points):
        minor_distance = -1
        for i1 in range(num_centroids):
            dx = arrayP[i0, 0] - arrayC[i1, 0]
            dy = arrayP[i0, 1] - arrayC[i1, 1]
            my_distance = numpy.sqrt(dx * dx + dy * dy)
            if minor_distance > my_distance or minor_distance == -1:
                minor_distance = my_distance
                arrayPcluster[i0] = i1
    return arrayPcluster


# assign points to cluster
@__njit
def calCentroidsSum(
    arrayP, arrayPcluster, arrayCsum, arrayCnumpoint, num_points, num_centroids
):
    # parallel for loop
    for i in numba.prange(num_centroids):
        arrayCsum[i, 0] = 0
        arrayCsum[i, 1] = 0
        arrayCnumpoint[i] = 0

    for i in range(num_points):
        ci = arrayPcluster[i]
        arrayCsum[ci, 0] += arrayP[i, 0]
        arrayCsum[ci, 1] += arrayP[i, 1]
        arrayCnumpoint[ci] += 1

    return arrayCsum, arrayCnumpoint


# update the centriods array after computation
@__njit
def updateCentroids(arrayC, arrayCsum, arrayCnumpoint, num_centroids):
    for i in numba.prange(num_centroids):
        arrayC[i, 0] = arrayCsum[i, 0] / arrayCnumpoint[i]
        arrayC[i, 1] = arrayCsum[i, 1] / arrayCnumpoint[i]


def kmeans(
    arrayP, arrayPcluster, arrayC, arrayCsum, arrayCnumpoint, num_points, num_centroids
):

    for i in range(ITERATIONS):
        groupByCluster(arrayP, arrayPcluster, arrayC, num_points, num_centroids)

        calCentroidsSum(
            arrayP, arrayPcluster, arrayCsum, arrayCnumpoint, num_points, num_centroids
        )

        updateCentroids(arrayC, arrayCsum, arrayCnumpoint, num_centroids)

    return arrayC, arrayCsum, arrayCnumpoint


def run_kmeans(
    arrayP,
    arrayPclusters,
    arrayC,
    arrayCsum,
    arrayCnumpoint,
    NUMBER_OF_POINTS,
    NUMBER_OF_CENTROIDS,
):

    with dpctl.device_context(get_device_selector(is_gpu=True)):
        for i in range(REPEAT):
            for i1 in range(NUMBER_OF_CENTROIDS):
                arrayC[i1, 0] = arrayP[i1, 0]
                arrayC[i1, 1] = arrayP[i1, 1]

            arrayC, arrayCsum, arrayCnumpoint = kmeans(
                arrayP,
                arrayPclusters,
                arrayC,
                arrayCsum,
                arrayCnumpoint,
                NUMBER_OF_POINTS,
                NUMBER_OF_CENTROIDS,
            )


base_kmeans.run("Kmeans Numba", run_kmeans)
